File: app/Services/Polling/RouterOsWireless.php
<?php

namespace App\Services\Polling;

use App\Models\Device;
use App\Services\RouterOs\RouterOsConnection;
use Illuminate\Support\Facades\Cache;
use Throwable;

/**
 * Wireless RF over the RouterOS API, for every MikroTik wireless stack:
 *
 *   wifi       /interface/wifi/registration-table        RouterOS 7.13+ (wifi-qcom / wifi-qcom-ac)
 *   wifiwave2  /interface/wifiwave2/registration-table   RouterOS 7.12 and older, same menu renamed
 *   wireless   /interface/wireless/registration-table    the legacy wireless package (v6 and v7)
 *   capsman    /caps-man/registration-table              legacy CAPsMAN controller (wireless package)
 *
 * Which menus a board has is probed once and cached per device + os_version (see stacks()), so a
 * normal poll only asks for the tables that exist. More than one can exist at once: from 7.13 the
 * wifi CAPsMAN controller ships in the base routeros package, so every 7.13+ router has an (empty)
 * /interface/wifi menu, and a hAP ac2 on 7.x with legacy radios has that plus /interface/wireless.
 * That's why we don't stop at the first menu that answers - we read every one that exists and
 * merge the rows. wifi and wifiwave2 never both exist (it's a rename), so wifiwave2 is only tried
 * when wifi isn't there, and caps-man only when the legacy wireless menu is.
 *
 * CAPsMAN: on a controller the registration table lists the stations of every CAP it manages
 * (wifi CAPsMAN puts the remote cap interfaces in the same /interface/wifi table, legacy CAPsMAN
 * has its own /caps-man table). We attribute all of them to the controller - it's the table the
 * controller actually has, and splitting per CAP would need a cap interface -> device mapping we
 * can't do reliably. Rows are de-duplicated by MAC across the merged tables so one station is
 * counted once even if it shows up in two of them mid roam.
 *
 * No rows at all -> everything null (not 0 clients), same as before. Every 7.13+ router has the
 * empty wifi table, reporting 0 there would make every wired router look like an AP.
 */
class RouterOsWireless
{
    /** Registration table menus, keyed by our stack name, in probe order. */
    public const MENUS = [
        'wifi' => '/interface/wifi/registration-table/print',
        'wifiwave2' => '/interface/wifiwave2/registration-table/print',
        'wireless' => '/interface/wireless/registration-table/print',
        'capsman' => '/caps-man/registration-table/print',
    ];

    /** How long a probe result is trusted. os_version is in the key, so an upgrade re-probes anyway. */
    private const CACHE_TTL = 6 * 3600;

    /**
     * Read and summarise RF for a device over an open connection. Never throws - a missing menu or
     * a failed query just means less (or no) RF this poll.
     *
     * @return array{signal:?float, snr:?float, ccq:?float, clients:?int}
     */
    public static function read(RouterOsConnection $conn, Device $device): array
    {
        $key = self::cacheKey($device);
        $cached = Cache::get($key);
        $rowsByStack = [];

        if (is_array($cached)) {
            foreach ($cached as $stack) {
                if (! isset(self::MENUS[$stack])) {
                    continue;
                }
                [$state, $rows] = self::query($conn, self::MENUS[$stack]);
                if ($state === 'missing') {
                    // The package went away (or got swapped) without a version change - probe
                    // again next poll rather than trapping on it forever.
                    Cache::forget($key);
                } elseif ($state === 'ok') {
                    $rowsByStack[$stack] = $rows;
                }
            }

            return self::summarise($rowsByStack);
        }

        $present = [];
        $complete = true;
        $try = static function (string $stack) use ($conn, &$present, &$complete, &$rowsByStack): bool {
            [$state, $rows] = self::query($conn, self::MENUS[$stack]);
            if ($state === 'ok') {
                $present[] = $stack;
                $rowsByStack[$stack] = $rows;

                return true;
            }
            if ($state === 'error') {
                $complete = false; // timeout or similar - don't cache a half answer
            }

            return false;
        };

        if (! $try('wifi')) {
            $try('wifiwave2');
        }
        if ($try('wireless')) {
            $try('capsman');
        }

        if ($complete) {
            Cache::put($key, $present, self::CACHE_TTL);
        }

        return self::summarise($rowsByStack);
    }

    public static function cacheKey(Device $device): string
    {
        return 'routeros:wl-stacks:'.$device->id.':'.($device->os_version ?? '');
    }

    /**
     * One registration table print. 'missing' = the menu doesn't exist on this board ("no such
     * command" / "no such command prefix"), 'error' = anything else went wrong.
     *
     * The PHP API lib doesn't throw on a !trap, it hands the trap's =message= back as if it were a
     * row, so a lone row carrying just a message is a trap too.
     *
     * @return array{0: 'ok'|'missing'|'error', 1: list<array<string, string>>}
     */
    private static function query(RouterOsConnection $conn, string $command): array
    {
        try {
            $rows = $conn->query($command);
        } catch (Throwable $e) {
            return [self::isMissingMenu($e->getMessage()) ? 'missing' : 'error', []];
        }

        if (count($rows) === 1 && isset($rows[0]['message']) && ! isset($rows[0]['mac-address'])) {
            return [self::isMissingMenu((string) $rows[0]['message']) ? 'missing' : 'error', []];
        }

        return ['ok', array_values($rows)];
    }

    private static function isMissingMenu(string $message): bool
    {
        return str_contains(strtolower($message), 'no such command');
    }

    /**
     * Rows from each stack -> the device's RF. Clients is the (MAC de-duplicated) row count,
     * signal / SNR / CCQ the average across the rows that have them.
     *
     * Field names per stack:
     *   wireless   signal-strength ("-65dBm@6Mbps"), signal-to-noise, tx-ccq
     *   wifi(wave2) signal (dBm, documented). No SNR and no CCQ in the documented table - 11ax has
     *              no CCQ, so it stays null rather than being made up. We still look at
     *              signal-strength / rx-signal and the per-chain signal-strength-chN / signal-chN
     *              in case a version names it differently (not documented, just defensive), and
     *              signal-to-noise if a later version ever adds it.
     *   capsman    rx-signal (dBm). No SNR, tx-ccq only if present.
     *
     * @param  array<string, list<array<string, string>>>  $rowsByStack
     * @return array{signal:?float, snr:?float, ccq:?float, clients:?int}
     */
    public static function summarise(array $rowsByStack): array
    {
        $signals = $snrs = $ccqs = [];
        $seen = [];
        $clients = 0;

        foreach ($rowsByStack as $stack => $rows) {
            foreach ($rows as $row) {
                $mac = strtoupper(trim((string) ($row['mac-address'] ?? '')));
                if ($mac !== '') {
                    if (isset($seen[$mac])) {
                        continue;
                    }
                    $seen[$mac] = true;
                }
                $clients++;

                $signal = match ($stack) {
                    'wireless' => self::firstNumber($row['signal-strength'] ?? null),
                    'capsman' => self::firstNumber($row['rx-signal'] ?? $row['signal-strength'] ?? null),
                    default => self::wifiSignal($row),
                };
                if ($signal !== null) {
                    $signals[] = $signal;
                }
                $snr = self::firstNumber($row['signal-to-noise'] ?? null);
                if ($snr !== null) {
                    $snrs[] = $snr;
                }
                // CCQ only exists on the legacy driver; wifi has none.
                if ($stack === 'wireless' || $stack === 'capsman') {
                    $ccq = self::firstNumber($row['tx-ccq'] ?? null);
                    if ($ccq !== null) {
                        $ccqs[] = $ccq;
                    }
                }
            }
        }

        if ($clients === 0) {
            return ['signal' => null, 'snr' => null, 'ccq' => null, 'clients' => null];
        }

        $avg = static fn (array $v): ?float => $v === [] ? null : round(array_sum($v) / count($v), 1);

        return [
            'signal' => $avg($signals),
            'snr' => $avg($snrs),
            'ccq' => $avg($ccqs),
            'clients' => $clients,
        ];
    }

    /**
     * A wifi / wifiwave2 row's signal. `signal` is the documented one. When only per-chain values
     * are there we take the strongest chain, which is closer to the combined figure than the mean
     * of the chains would be.
     *
     * @param  array<string, string>  $row
     */
    private static function wifiSignal(array $row): ?float
    {
        foreach (['signal', 'signal-strength', 'rx-signal'] as $field) {
            $v = self::firstNumber($row[$field] ?? null);
            if ($v !== null) {
                return $v;
            }
        }

        $chains = [];
        foreach ($row as $field => $value) {
            if (preg_match('/^(signal|signal-strength|rx-signal)-ch\d+$/', (string) $field) === 1) {
                $v = self::firstNumber($value);
                if ($v !== null) {
                    $chains[] = $v;
                }
            }
        }

        return $chains === [] ? null : max($chains);
    }

    /** First signed/decimal number in a value (e.g. "-65dBm@6Mbps" -> -65.0), or null. */
    private static function firstNumber(mixed $value): ?float
    {
        if ($value === null) {
            return null;
        }

        return preg_match('/-?\d+(\.\d+)?/', (string) $value, $m) === 1 ? (float) $m[0] : null;
    }
}

File: agent/internal/poll/routeros_wireless.go
package poll

import (
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/AthenaNetworks/mymate/agent/internal/proto"
)

// Wireless RF over the RouterOS API for every MikroTik wireless stack, the same as the central
// RouterOsWireless (see there for the long version):
//
//	wifi       /interface/wifi/registration-table        RouterOS 7.13+
//	wifiwave2  /interface/wifiwave2/registration-table   7.12 and older, same menu renamed
//	wireless   /interface/wireless/registration-table    legacy wireless package
//	capsman    /caps-man/registration-table              legacy CAPsMAN controller
//
// Every menu that exists is read and the rows merged (a 7.13+ router always has the empty wifi
// menu from the base package, and can have legacy radios too). wifiwave2 is only tried when wifi
// isn't there, caps-man only when legacy wireless is. What a board has is probed once and kept per
// device + os_version for wlStackTTL, so a normal poll doesn't trap on missing menus every time.
//
// A CAPsMAN controller's table lists the stations of all its CAPs; they're all attributed to the
// controller, de-duplicated by MAC so one station is counted once.

var routerOSWirelessMenus = map[string]string{
	"wifi":      "/interface/wifi/registration-table/print",
	"wifiwave2": "/interface/wifiwave2/registration-table/print",
	"wireless":  "/interface/wireless/registration-table/print",
	"capsman":   "/caps-man/registration-table/print",
}

const wlStackTTL = 6 * time.Hour

// rosRun runs one print and returns its rows. The real one wraps a routeros client; tests fake it.
type rosRun func(cmd string) ([]map[string]string, error)

// wlStackCache remembers which registration table menus a device has.
type wlStackCache struct {
	mu sync.Mutex
	m  map[string]wlStackEntry
}

type wlStackEntry struct {
	stacks []string
	at     time.Time
}

func (c *wlStackCache) get(key string, now time.Time) ([]string, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	e, ok := c.m[key]
	if !ok || now.Sub(e.at) > wlStackTTL {
		return nil, false
	}
	return e.stacks, true
}

func (c *wlStackCache) put(key string, stacks []string, now time.Time) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.m == nil {
		c.m = map[string]wlStackEntry{}
	}
	c.m[key] = wlStackEntry{stacks: stacks, at: now}
}

func (c *wlStackCache) forget(key string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	delete(c.m, key)
}

// isMissingMenu is a !trap for a menu the board doesn't have ("no such command prefix").
// go-routeros returns a trap as an error, anything else (timeout, dropped conn) is a real error.
func isMissingMenu(err error) bool {
	return err != nil && strings.Contains(strings.ToLower(err.Error()), "no such command")
}

type stackRows struct {
	stack string
	rows  []map[string]string
}

// routerOSWirelessRead reads RF for one device. Never fails - a missing menu or a failed print
// just means less (or no) RF this poll.
func (p *Poller) routerOSWirelessRead(t proto.RouterOSTarget, run rosRun, now time.Time) proto.Wireless {
	key := fmt.Sprintf("%d|%s", t.DeviceID, t.OSVersion)
	var got []stackRows

	if stacks, ok := p.wl.get(key, now); ok {
		for _, s := range stacks {
			rows, err := run(routerOSWirelessMenus[s])
			if isMissingMenu(err) {
				p.wl.forget(key) // package removed without a version change, probe again next time
				continue
			}
			if err == nil {
				got = append(got, stackRows{s, rows})
			}
		}
		return routerOSWirelessStacks(got)
	}

	present := []string{}
	complete := true
	try := func(s string) bool {
		rows, err := run(routerOSWirelessMenus[s])
		if err == nil {
			present = append(present, s)
			got = append(got, stackRows{s, rows})
			return true
		}
		if !isMissingMenu(err) {
			complete = false // don't cache a half answer
		}
		return false
	}
	if !try("wifi") {
		try("wifiwave2")
	}
	if try("wireless") {
		try("capsman")
	}
	if complete {
		p.wl.put(key, present, now)
	}
	return routerOSWirelessStacks(got)
}

// routerOSWireless turns legacy wireless registration-table rows into RF.
func routerOSWireless(rows []map[string]string) proto.Wireless {
	return routerOSWirelessStacks([]stackRows{{"wireless", rows}})
}

var signalChain = regexp.MustCompile(`^(signal|signal-strength|rx-signal)-ch\d+$`)

// routerOSWirelessStacks is the merged RF, same as the central RouterOsWireless::summarise: one
// row per associated station (an AP sees its clients, a station sees its AP), so clients is the
// row count (de-duplicated by MAC) and signal / SNR / CCQ the average across the rows that have
// them. Values like "-65dBm@6Mbps" read as their leading number. No rows, no RF - not 0 clients,
// since every 7.13+ router has an empty wifi table.
//
// Fields: legacy wireless signal-strength / signal-to-noise / tx-ccq. wifi and wifiwave2 document
// only `signal` (dBm), no SNR and no CCQ, so CCQ is never read there. signal-strength / rx-signal
// and the per-chain -chN variants are a defensive guess in case a version names it differently,
// the strongest chain standing in for the combined value. caps-man gives rx-signal.
func routerOSWirelessStacks(stacks []stackRows) proto.Wireless {
	var signals, snrs, ccqs []float64
	seen := map[string]bool{}
	clients := 0
	num := func(v string) (float64, bool) {
		m := leadingNumber.FindString(v)
		if m == "" {
			return 0, false
		}
		f, err := strconv.ParseFloat(m, 64)
		return f, err == nil
	}
	for _, s := range stacks {
		for _, row := range s.rows {
			if mac := strings.ToUpper(strings.TrimSpace(row["mac-address"])); mac != "" {
				if seen[mac] {
					continue
				}
				seen[mac] = true
			}
			clients++

			switch s.stack {
			case "wireless":
				if f, ok := num(row["signal-strength"]); ok {
					signals = append(signals, f)
				}
			case "capsman":
				v := row["rx-signal"]
				if v == "" {
					v = row["signal-strength"]
				}
				if f, ok := num(v); ok {
					signals = append(signals, f)
				}
			default:
				if f, ok := wifiSignal(row, num); ok {
					signals = append(signals, f)
				}
			}
			if f, ok := num(row["signal-to-noise"]); ok {
				snrs = append(snrs, f)
			}
			if s.stack == "wireless" || s.stack == "capsman" {
				if f, ok := num(row["tx-ccq"]); ok {
					ccqs = append(ccqs, f)
				}
			}
		}
	}
	if clients == 0 {
		return proto.Wireless{}
	}
	avg1 := func(v []float64) *float64 {
		if len(v) == 0 {
			return nil
		}
		r := math.Round(sum(v)/float64(len(v))*10) / 10
		return &r
	}
	return proto.Wireless{SignalDbm: avg1(signals), SnrDb: avg1(snrs), CcqPct: clampPct(avg1(ccqs)), Clients: &clients}
}

func wifiSignal(row map[string]string, num func(string) (float64, bool)) (float64, bool) {
	for _, f := range []string{"signal", "signal-strength", "rx-signal"} {
		if v, ok := num(row[f]); ok {
			return v, true
		}
	}
	best, found := 0.0, false
	for k, v := range row {
		if !signalChain.MatchString(k) {
			continue
		}
		if f, ok := num(v); ok && (!found || f > best) {
			best, found = f, true
		}
	}
	return best, found
}
